modulemd/v2/modulemd-defaults-v1.c

```diff
@@ -46,6 +47,105 @@ G_DEFINE_TYPE (ModulemdDefaultsV1,
                MODULEMD_TYPE_DEFAULTS)
 
 
+static gboolean
+modulemd_defaults_v1_equals (ModulemdDefaults *self_1, ModulemdDefaults *self_2, GError **error)
+{
+  ModulemdDefaultsV1 *v1_self_1 = NULL;
+  ModulemdDefaultsV1 *v1_self_2 = NULL;
+  g_autoptr (GError) nested_error = FALSE;
+  g_autoptr (GPtrArray) profile_defaults_1 = NULL;
+  g_autoptr (GPtrArray) profile_defaults_2 = NULL;
+  g_autoptr (GPtrArray) intent_default_profiles_1 = NULL;
+  g_autoptr (GPtrArray) intent_default_profiles_2 = NULL;
+  gchar *key_1 = NULL;
+  gchar *key_2 = NULL;
+  GHashTable * table_1 = NULL;
+  GHashTable * table_2 = NULL;
+  int i = 0;
+
+  g_return_val_if_fail (MODULEMD_IS_DEFAULTS_V1 (self_1), FALSE);
+  v1_self_1 = MODULEMD_DEFAULTS_V1 (self_1);  
+  g_return_val_if_fail (MODULEMD_IS_DEFAULTS_V1 (self_2), FALSE);
+  v1_self_2 = MODULEMD_DEFAULTS_V1 (self_2);  
+
+  if (!MODULEMD_DEFAULTS_CLASS (modulemd_defaults_v1_parent_class)
+             ->equals (self_1, self_2, error))
+    {
+      g_propagate_error (error, g_steal_pointer (&nested_error));
+      return FALSE;
+    }
+
+  if(g_strcmp0 (v1_self_1->default_stream, v1_self_2->default_stream) != 0)
+    {
+      return FALSE;
+    } 
+
+  //Check profile_defaults: size, keys, values
+  if (g_hash_table_size (v1_self_1->profile_defaults) != g_hash_table_size(v1_self_2->profile_defaults))
+    {
+      return FALSE;
+    }
+  profile_defaults_1 = modulemd_ordered_str_keys (v1_self_1->profile_defaults, modulemd_strcmp_sort);
+  profile_defaults_2 = modulemd_ordered_str_keys (v1_self_2->profile_defaults, modulemd_strcmp_sort);
+
+  for (i = 0; i < profile_defaults_1->len; i++)
+    {
+      key_1 = g_ptr_array_index (profile_defaults_1, i);
+      key_2 = g_ptr_array_index (profile_defaults_2, i);
+
+      if(g_strcmp0 (key_1, key_2) != 0)
+        {
+          return FALSE;
+        } 
+      }
+
+  for (i = 0; i < profile_defaults_1->len; i++)
```

Comment:
This implementation is correct, but I'm wondering if we should make a generic hash-table comparison function that just takes a function pointer for a comparison function to handle the comparison of the values. This would be similar to how `modulemd_ordered_str_keys()` works; we'd want to take a `GEqualFunc ()` pointer and then implement it for each value type.

This could then be reused for any hash table equality check we might need to write, such as the intent_default_profiles just below here, as well as probably for the ModuleStream.equals() later.

Reply:
So I should make a function that first checks the size, then keys, then its corresponding values for equality. Basically wrap all of this in 1 function like:

```
if (!modulemd_hash_table_are_equal(self_1->profile_default, self_2->profile_default))
{
      return FALSE;
}

```

Also, what file will that change be in? I see only function declarations in `private/modulemd-util.h`

Reply:
Yes, almost exactly. I'd probably call the function `modulemd_hash_table_equals()` and it will need to take three arguments (not two): `modulemd_hash_table_equals (self_1->profile_default, self_2->profile-default, profile_default_equals)`

And `profile_default_equals()` will be a function of type [`GEqualFunc()`](https://developer.gnome.org/glib/stable/glib-Hash-Tables.html#GEqualFunc), which you will use after comparing the keys to compare the values for each key. You will need to implement a unique equality function for any type of value you need to deal with.